File: sdk/typescript/src/result.ts
import { AegisExecutionError, AegisVerificationError } from "./errors.js";
import { Receipt } from "./receipt.js";
import { ProofBundle } from "./types.js";
import { ReceiptVerification, ReceiptVerifier } from "./verifier.js";

export interface ExecutionResponse {
  stdout?: string;
  stderr?: string;
  exit_code?: number;
  execution_id?: string;
  duration_ms?: number;
  exit_reason?: string;
  error?: string;
  output_truncated?: boolean;
  proof_dir?: string;
  receipt_path?: string;
  receipt_public_key_path?: string;
  receipt_summary_path?: string;
  [key: string]: unknown;
}

export class ExecutionResult {
  readonly stdout: string;
  readonly stderr: string;
  readonly exitCode: number;
  readonly executionId: string;
  readonly durationMs: number;
  readonly exitReason?: string;
  readonly executionError?: string;
  readonly outputTruncated: boolean;
  readonly proofDir?: string;
  readonly receiptPath?: string;
  readonly receiptPublicKeyPath?: string;
  readonly receiptSummaryPath?: string;
  readonly rawResponse: Record<string, unknown>;
  private readonly verifier?: ReceiptVerifier;
  private receiptCache?: Receipt | null;

  constructor(input: {
    stdout: string;
    stderr: string;
    exitCode: number;
    executionId: string;
    durationMs: number;
    exitReason?: string;
    executionError?: string;
    outputTruncated: boolean;
    proofDir?: string;
    receiptPath?: string;
    receiptPublicKeyPath?: string;
    receiptSummaryPath?: string;
    rawResponse: Record<string, unknown>;
    verifier?: ReceiptVerifier;
  }) {
    this.stdout = input.stdout;
    this.stderr = input.stderr;
    this.exitCode = input.exitCode;
    this.executionId = input.executionId;
    this.durationMs = input.durationMs;
    this.exitReason = input.exitReason;
    this.executionError = input.executionError;
    this.outputTruncated = input.outputTruncated;
    this.proofDir = input.proofDir;
    this.receiptPath = input.receiptPath;
    this.receiptPublicKeyPath = input.receiptPublicKeyPath;
    this.receiptSummaryPath = input.receiptSummaryPath;
    this.rawResponse = input.rawResponse;
    this.verifier = input.verifier;
  }

  static fromPayload(payload: ExecutionResponse, verifier?: ReceiptVerifier): ExecutionResult {
    return new ExecutionResult({
      stdout: typeof payload.stdout === "string" ? payload.stdout : "",
      stderr: typeof payload.stderr === "string" ? payload.stderr : "",
      exitCode: typeof payload.exit_code === "number" ? payload.exit_code : 0,
      executionId: typeof payload.execution_id === "string" ? payload.execution_id : "",
      durationMs: typeof payload.duration_ms === "number" ? payload.duration_ms : 0,
      exitReason: typeof payload.exit_reason === "string" ? payload.exit_reason : undefined,
      executionError: typeof payload.error === "string" ? payload.error : undefined,
      outputTruncated: Boolean(payload.output_truncated),
      proofDir: typeof payload.proof_dir === "string" ? payload.proof_dir : undefined,
      receiptPath: typeof payload.receipt_path === "string" ? payload.receipt_path : undefined,
      receiptPublicKeyPath: typeof payload.receipt_public_key_path === "string" ? payload.receipt_public_key_path : undefined,
      receiptSummaryPath: typeof payload.receipt_summary_path === "string" ? payload.receipt_summary_path : undefined,
      rawResponse: { ...payload },
      verifier,
    });
  }

  get proofBundle(): ProofBundle {
    return new ProofBundle({
      proofDir: this.proofDir,
      receiptPath: this.receiptPath,
      receiptPublicKeyPath: this.receiptPublicKeyPath,
      receiptSummaryPath: this.receiptSummaryPath,
    });
  }

  get executionFailed(): boolean {
    if (this.executionError) return true;
    if (this.exitCode !== 0) return true;
    return ![undefined, "", "completed"].includes(this.exitReason);
  }

  get ok(): boolean {
    return !this.executionFailed;
  }

  get error(): string | undefined {
    if (this.executionError) return this.executionError;
    if (this.exitCode !== 0) return `process exited with code ${this.exitCode}`;
    if (this.exitReason && this.exitReason !== "completed") return this.exitReason;
    return undefined;
  }

  get receipt(): Receipt | undefined {
    if (this.receiptCache !== undefined) return this.receiptCache ?? undefined;
    if (!this.receiptPath) {
      this.receiptCache = null;
      return undefined;
    }
    try {
      this.receiptCache = Receipt.load(this.receiptPath, {
        publicKeyPath: this.receiptPublicKeyPath,
        summaryPath: this.receiptSummaryPath,
        verifier: this.verifier,
      });
    } catch {
      this.receiptCache = null;
    }
    return this.receiptCache ?? undefined;
  }

  requireReceipt(): Receipt {
    const receipt = this.receipt;
    if (!receipt) throw new AegisVerificationError("execution result does not include a readable receipt");
    return receipt;
  }

  async verifyReceipt(): Promise<ReceiptVerification> {
    return this.requireReceipt().verify();
  }

  throwIfExecutionFailed(): void {
    if (!this.executionFailed) return;
    throw new AegisExecutionError(this.error ?? "execution failed", {
      executionId: this.executionId,
      exitCode: this.exitCode,
      exitReason: this.exitReason,
      rawResponse: this.rawResponse,
    });
  }
}

File: sdk/python/aegis/result.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .errors import AegisExecutionError, AegisVerificationError
from .receipt import Receipt
from .types import ProofBundle
from .verifier import ReceiptVerification, ReceiptVerifier


@dataclass(slots=True)
class ExecutionResult:
    stdout: str
    stderr: str
    exit_code: int
    execution_id: str
    duration_ms: int = 0
    exit_reason: str | None = None
    execution_error: str | None = None
    output_truncated: bool = False
    proof_dir: str | None = None
    receipt_path: str | None = None
    receipt_public_key_path: str | None = None
    receipt_summary_path: str | None = None
    raw_response: dict[str, Any] = field(default_factory=dict)
    _verifier: ReceiptVerifier | None = field(default=None, repr=False, compare=False)
    _receipt_cache: Receipt | None = field(default=None, init=False, repr=False, compare=False)

    @classmethod
    def from_dict(cls, payload: dict[str, Any], verifier: ReceiptVerifier | None = None) -> ExecutionResult:
        return cls(
            stdout=str(payload.get("stdout", "")),
            stderr=str(payload.get("stderr", "")),
            exit_code=int(payload.get("exit_code", 0) or 0),
            execution_id=str(payload.get("execution_id", "")),
            duration_ms=int(payload.get("duration_ms", 0) or 0),
            exit_reason=payload.get("exit_reason"),
            execution_error=payload.get("error"),
            output_truncated=bool(payload.get("output_truncated", False)),
            proof_dir=payload.get("proof_dir"),
            receipt_path=payload.get("receipt_path"),
            receipt_public_key_path=payload.get("receipt_public_key_path"),
            receipt_summary_path=payload.get("receipt_summary_path"),
            raw_response=dict(payload),
            _verifier=verifier,
        )

    @property
    def proof_bundle(self) -> ProofBundle:
        return ProofBundle(
            proof_dir=self.proof_dir,
            receipt_path=self.receipt_path,
            receipt_public_key_path=self.receipt_public_key_path,
            receipt_summary_path=self.receipt_summary_path,
        )

    @property
    def ok(self) -> bool:
        return not self.execution_failed

    @property
    def execution_failed(self) -> bool:
        if self.execution_error:
            return True
        if self.exit_code != 0:
            return True
        return self.exit_reason not in {None, "", "completed"}

    @property
    def error(self) -> str | None:
        if self.execution_error:
            return self.execution_error
        if self.exit_code != 0:
            return f"process exited with code {self.exit_code}"
        if self.exit_reason not in {None, "", "completed"}:
            return self.exit_reason
        return None

    @property
    def receipt(self) -> Receipt | None:
        if self._receipt_cache is not None:
            return self._receipt_cache
        if not self.receipt_path:
            return None
        try:
            self._receipt_cache = Receipt.load(
                self.receipt_path,
                public_key_path=self.receipt_public_key_path,
                summary_path=self.receipt_summary_path,
                verifier=self._verifier,
            )
        except FileNotFoundError:
            return None
        return self._receipt_cache

    def require_receipt(self) -> Receipt:
        receipt = self.receipt
        if receipt is None:
            raise AegisVerificationError("execution result does not include a readable receipt")
        return receipt

    def verify_receipt(self) -> ReceiptVerification:
        return self.require_receipt().verify()

    def raise_for_execution_error(self) -> None:
        if not self.execution_failed:
            return
        raise AegisExecutionError(
            message=self.error or "execution failed",
            execution_id=self.execution_id,
            exit_code=self.exit_code,
            exit_reason=self.exit_reason,
            raw_response=self.raw_response,
        )
